From: NeilBrown <neilb@suse.de>
Date: Wed, 20 Apr 2011 15:38:18 +1000
Subject: [PATCH] md: Fix dev_sectors on takeover from raid0 to raid4/5
Git-commit: 3b71bd9337b404baab5c894e066be6b6bf51b1c3
Patch-mainline: v2.6.39-rc5

A raid0 array doesn't set 'dev_sectors' as each device might
contribute a different number of sectors.
So when converting to a RAID4 or RAID5 we need to set dev_sectors
as they need the number.
We have already verified that in fact all devices do contribute
the same number of sectors, so use that number.

Signed-off-by: NeilBrown <neilb@suse.de>
---
 drivers/md/raid5.c |    4 ++++
 1 file changed, 4 insertions(+)

--- linux-2.6.32-SLE11-SP1.orig/drivers/md/raid5.c
+++ linux-2.6.32-SLE11-SP1/drivers/md/raid5.c
@@ -5723,6 +5723,7 @@ static void raid5_quiesce(mddev_t *mddev
 static void *raid45_takeover_raid0(mddev_t *mddev, int level)
 {
 	struct raid0_private_data *raid0_priv = mddev->private;
+	unsigned long long sectors;
 
 	/* for raid0 takeover only one zone is supported */
 	if (raid0_priv->nr_strip_zones > 1) {
@@ -5730,6 +5731,9 @@ static void *raid45_takeover_raid0(mddev
 		return ERR_PTR(-EINVAL);
 	}
 
+	sectors = raid0_priv->strip_zone[0].zone_end;
+	sector_div(sectors, raid0_priv->strip_zone[0].nb_dev);
+	mddev->dev_sectors = sectors;
 	mddev->new_level = level;
 	mddev->new_layout = ALGORITHM_PARITY_N;
 	mddev->new_chunk_sectors = mddev->chunk_sectors;
